asciidtype/asciidtype/src/umath.c

```diff
@@ -9,4 +9,229 @@
 #include "numpy/ufuncobject.h"
 
 #include "dtype.h"
+#include "string.h"
 #include "umath.h"
+
+static int
+ascii_add_strided_loop(PyArrayMethod_Context *context, char *const data[],
+                       npy_intp const dimensions[], npy_intp const strides[],
+                       NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in1_size = ((ASCIIDTypeObject *)descrs[0])->size;
+    long in2_size = ((ASCIIDTypeObject *)descrs[1])->size;
+
+    npy_intp N = dimensions[0];
+    char *in1 = data[0], *in2 = data[1], *out = data[2];
+    npy_intp in1_stride = strides[0], in2_stride = strides[1],
+             out_stride = strides[2];
+
+    while (N--) {
+        strncpy(out, in1, in1_size);
+        strncpy(out + in1_size, in2, in2_size);
+        in1 += in1_stride;
+        in2 += in2_stride;
+        out += out_stride;
+    }
+
+    return 0;
+}
+
+static NPY_CASTING
+ascii_add_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                              PyArray_DTypeMeta *dtypes[],
+                              PyArray_Descr *given_descrs[],
+                              PyArray_Descr *loop_descrs[],
+                              npy_intp *NPY_UNUSED(view_offset))
+{
+    long op1_size = ((ASCIIDTypeObject *)given_descrs[0])->size;
+    long op2_size = ((ASCIIDTypeObject *)given_descrs[1])->size;
```

Comment:
One of these can be `NULL` for `np.sum([...])`.  Otherwise things are fine if you return almost anything:  NumPy will complain automatically for `sum()` if (first) input and output dtypes don't match.

Reply:
Good point! I hadn't thought about reductions at all up until now.

When I do `np.sum` now, I get that neither is NULL. There's [a comment](https://github.com/numpy/numpy/blob/5e0ed03b6b7a6ff05843d846eedac730f20decb7/numpy/core/src/umath/ufunc_object.c#L2795-L2800) that the first argument should be eventually though. The way it's set up right now I don't see a way for me to tell the difference from inside `resolve_descriptors` between a reduction and a regular ufunc operation. Am I missing something there?

Reply:
Currently the reduction fails with the following error:

```
TypeError: the resolved dtypes are not compatible with add.reduce. Resolved (ASCIIDType(5), ASCIIDType(5), ASCIIDType(10))
```

I guess this is another place where ideally we'd be doing value-based promotion, but I guess for practicality we could return the input dtype and just truncate. However, I'd need for one of the `given_descrs` to be NULL to enable that I think,

Reply:
Ah OK, so it can't actually happen right now.  The error is what I referred to and I think it is fine for now.  But of course you could calculate a "correct" length here if you knew how many elements are being summed (for strings that would be weird, but for units the count is really enough).

Reply:
Yeah, agreed, getting the number of elements or array shape inside `resolve_descriptors` would be nice for reductions.